text/3476-expose-fn-type.md

```diff
@@ -0,0 +1,164 @@
+
+# Motivation
+[motivation]: #motivation
+
+I was trying to make something similar to bevy's system functions. And for safety reasons, they check for conflicts between SystemParams, so that a function requiring `Res<A>` and `ResMut<A>` [panic](https://github.com/bevyengine/bevy/blob/main/crates/bevy_ecs/src/system/system_param.rs#L421).
+
+Then after I heard about axum's [`#[debug_handler]`](https://docs.rs/axum/latest/axum/attr.debug_handler.html) I wanted to do something similar to my copy of bevy systems, so that I get compile time errors when there is a conflict. I wanted even more, I wanted to force the user to mark the function with a specific proc attribute macro in order to make it possible to pass it into my code and call itself a system.
+
+For that, I would need to mark the type behind the function, for example, with a trait.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As we all know, you can refer to a struct by its name and for example implement a trait
+```rust
+struct Timmy;
+impl Person for Timmy {
+    fn greet() {
+        println!("Hey it's me, Timmy!");
+    }
+}
+```
+When we want to target a specific function for a trait implementation, we somehow need to get to the type behind it. That is being done with the `fn` keyword as follows
+```rust
+fn my_function() {}
+impl MyTrait for fn my_function {
+    /* ... */
+}
+```
+---
+For a better understanding, imagine you have a struct like this:
+```rust
+struct FnContainer<F: Fn()> {
+    inner: F,
+}
+fn goods() { }
+
+let contained_goods = FnContainer {
+    inner: goods
+};
+```
+Here, we make a `FnContainer` which can hold every function with the signature `() -> ()` via generics.
+But what about explicitly designing the `FnContainer` for a specific function, just like the compiler does when resolving the generics. This will work the same as with the trait impl from above:
+```rust
+struct GoodsContainer {
+    inner: fn goods,
+}
+fn goods() {}
+
+let contained_goods = GoodsContainer {
+    inner: goods,
+}
+```
+---
+A function with a more complex signature, like with parameters, modifiers or a return type, is still just referenced by its name, because it's already unique
+```rust
+async fn request_name(id: PersonID) -> String { .. }
+
+impl Requestable for fn request_name {
+    /* ... */
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+As described in the **Guide-level explanation**, with the syntax `fn <fn_path>`, we can reference the type behind the named function.
+
+When the function is for example in a different mod, it should be referenced by its path
+```rust
+mod sub {
+    fn sub_mod_fn() { .. }
+}
+impl fn sub::sub_mod_fn {
+    /* ... */
+}
+```
+
+It should be also possible to get the type of functions inside impl blocks:
+
+```rust
+struct MyStruct;
+impl MyStruct {
+    fn new() -> Self { Self }
+}
+impl fn MyStruct::new {
```

Comment:
Do we use turbofish syntax if `MyStruct` takes a generic parameter? e.g.

```rust
struct MyStruct<T>(T);
```

Do we write `fn MyStruct<T>::new` or `fn MyStruct::<T>::new`? If we choose the latter, would it cause any inconsistency from the `fn send<T>` syntax below (as opposed to `fn send::<T>`)?

Reply:
I'll look into that. I also see another problem here: Imagine this
```rust
struct MyStruct<T> {
    pub fn new(x: T) -> Self {
        // ...
    }
}
```
is it
`impl<T> fn MyStruct[::]<T>::new`
or is it
`impl<T> fn MyStruct::new[::]<T>`
?
what do you think? is there a prefered situation for this in rust or should we maybe allow both? @SOF3

Reply:
definitely the former. `MyStruct::new<T>` doesn't make sense because the type parameter is on MyStruct not the associated function. This would go wrong if `new` itself also accepts type parameters.

As for whether to use turbofish, I guess this depends on how the compiler parses this expression. I'm not a parser expert, so this part needs some input from the compiler team.

Reply:
@SOF3 So I made some tests and would say that it's more "rusty" when we do `MyStruct::<T>::new` instead of `MyStruct<T>::new` (https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=316deb57a2d9552d8284a35fb56db2a0)

What do you think?

Reply:
Yep, but since we are in type context not expression context here, turbofish is probably unnecessary. The turbofish syntax is required for expressions only because of ambiguity with the less-than operator (`<`), but we don't have such ambiguity if we specify that `fn` is always followed by a path instead of an expression.

Of course, you might also want an expression there if it is a generic `typeof` operator, but this is not the case for the scope of this RFC.